### ISSUE TITLE
Show on-disk file size in addition to 'added to repo' in backup summary

### ISSUE THREAD

**Issue by OJFord** [restic/restic#3431]
Output of `restic version`
--------------------------

restic 0.12.0 compiled with go1.15.8 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

A successful `restic backup` outputs a summary that ends with something like:
> Added to the repo: 12.345 MiB

It'd be nice if this showed both the 'real' size of added/changed/removed files, as on local disk, in addition to the amount 'added to the repo', which could be significantly less.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

It'd just be a nice at-a-glance summary of how well the added files, in particular, were covered by extant blocks.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Just a suggestion from yet another happy user.

**Comment by calr0x** [comment]
I'd love to see this as well!
Thanks for the great project!

**Comment by OJFord** [comment]
I'm really confused now - I thought I was experiencing this the other way around, but vaguely recalled this issue (didn't think I'd opened it! very short memory..).

Either way I think it'd be useful to show both, but which is it currently, what does 'Added to repo' mean:
1. The cumulative file size of new/modified/deleted files in the snapshot; or
2. The additional remote storage required by the snapshot, after block-wise de-duplication, etc.

? I think it could be clearer which it means even if it's not wanted (or not possible/too difficult) to add the 'other one'.

**Comment by MichaelEischer** [comment]
Restic reports how much deduplicated data was uploaded to the repository, without accounting for overheads due to encryption and other repository metadata. That is its roughly option 2.

**Comment by OJFord** [comment]
Thanks @MichaelEischer. I just noticed though, after `restic backup`:
> Added to the repo: 2.115 GiB

and then `restic diff <parent> <new>`:
>   Added:   2.115 GiB
>   Removed: 1.727 GiB

is the reason 'added to the repo' is not net of the 'removed' figure that all of what was removed is still referenced by at least the parent snapshot, so we added 2.115GiB worth of new (de-duplicated) blocks, removed references to 1.727GiB worth of old ones, but until a prune, `restic backup` itself will only ever _add_, and always equal to the 'added' figure in the resulting diff?

Many thanks.

**Comment by MichaelEischer** [comment]
> > Added:   2.115 GiB
> > Removed: 1.727 GiB
> 
> is the reason 'added to the repo' is not net of the 'removed' figure that all of what was removed is still referenced by at least the parent snapshot, so we added 2.115GiB worth of new (de-duplicated) blocks, removed references to 1.727GiB worth of old ones, but until a prune, `restic backup` itself will only ever _add_, and always equal to the 'added' figure in the resulting diff?

The added and removed numbers of `backup` and `diff` are independent of each other. `backup` reports how much new data was uploaded. `diff` reports the change between two snapshots. The `added` numbers can match, but there's no need to. 

Just assume that you have a file "x" that has already been backed up in the past, but which you deleted in the meantime. That is "x" would not be included in the parent snapshot. Now, restore file "x". Then the new snapshot will include file "x" again. `backup` will deduplicate the file (as it still exists in an older snapshot) and therefore not add data to the repository. `diff` on the other hand will report that "x" has been added.

The `Removed` numbers show how much space could be freed by deleting the parent snapshot, under the assumption that the removed files are not referenced by any other snapshot. That is `Removed` is an upper boundary on how much space can be freed by deleting the parent snapshot.

A more reliable way to determine how much data can be freed by removing certain snapshots is to use `restic stats --mode raw-data <kept-snapshots>` to calculate how much space is required to store the snapshots that should be kept. Then subtract the repository size from that result.

**Comment by maltem-za** [comment]
Isn't the on-disk size covered by the "processed <num> files, <size> in <time>" output just before the hash of the new snapshot? Not sure if that was added after this issue was created...

**Comment by kjetilho** [comment]
> Isn't the on-disk size covered by the "processed <num> files, <size> in <time>" output just before the ID of the new snapshot? Not sure if that was added after this issue was created...

I don't think so, that is the total size of the filesystem in the backup, so will be the same for every snapshot of the file set.

Consider a fairly typical report at the completion of a job:

```
Files:      12 new,       3 changed, 7075229 unmodified
Dirs:        0 new,  216721 changed,       0 unmodified
```

I think the original reporter wants the total size of the 12 new files, and possibly also the total size of the 3 changed files - both behaviours make sense to me.  Possibly also the size of the changed directories, but probably not.  (I don't understand why dirs almost always change.)

This statistic will only make sense when the job has a parent as a reference to identify what files needs processing, without a parent it will be the total size of the backup snapshot.  Collecting this statistic has little impact on performance, it's just a single byte counter - although locking that counter to allow for concurrent threads can conceivably have an impact.

**Comment by maltem-za** [comment]
> I think the original reporter wants the total size of the 12 new files, and possibly also the total size of the 3 changed files - both behaviours make sense to me.

Yes, I think you're right. Somehow I got stuck on the "on disk" part and skipped that the wish was for a kind of delta.

> I don't think so, that is the total size of the filesystem in the backup, [...]

I had to think about what you meant here and would respectfully suggest a rephrasing for clarity:
I would say that it is the total size of the directories and files that were backed up.
(At least in my view, "filesystem" can be interpreted as meaning more than the directories and files that were backed up, and "in the backup" can be interpreted as "in the snapshot/repo", but there parent snapshots and deduplication come into play. In the end a snapshot does of course mirror what was backed up, so "in the snapshot" is probably an OK interpretation... nuance! That part is about source vs. destination for me though.)

> [...] so will be the same for every snapshot of the file set.

Only if we assume that nothing has changed in the file set.

Do you agree?
